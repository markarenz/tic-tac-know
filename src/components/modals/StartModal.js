import React from 'react';
import { TransitionWrap } from '../../components';
import PropTypes from 'prop-types';

const StartModal = ({ show, handleStartClick, handleResetClick }) => {
  return (
    <TransitionWrap show={show} className="text-center">
      <div className="text-center mt-16 inline-block">
        <div className="relative">
          <div className="animate-pingSlowSm absolute top-0 left-0 inline-flex h-full w-full rounded-lg bg-purple-800 opacity-100"></div>
          <div className="mb-8">
            <button
              data-testid="startmenu-play"
              className="relative px-16 py-4 text-4xl border-2 border-purple-300 text-white bg-gradient-to-t from-black to-purple-700 rounded-lg scale-100 hover:scale-[1.2] transition-scale font-bold duration-300"
              onClick={handleStartClick}
            >
              PLAY
            </button>
          </div>
        </div>
        <div className="mt-4">
          <button
            data-testid="startmenu-reset"
            className="relative px-4 py-2 text-lg border-2 border-purple-300 text-white bg-gradient-to-t from-black to-gray-500 rounded-lg scale-100 hover:scale-[1.2] transition-scale duration-300"
            onClick={handleResetClick}
          >
            RESET
          </button>
        </div>
      </div>
    </TransitionWrap>
  );
};

StartModal.propTypes = {
  show: PropTypes.bool.isRequired,
  handleStartClick: PropTypes.func.isRequired,
  handleResetClick: PropTypes.func.isRequired,
};

export default StartModal;
