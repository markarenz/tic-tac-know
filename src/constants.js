// STRINGS
const PLAYER = 'player';
const CPU = 'cpu';
const X = 'x';
const O = 'o';

const AI_LEVELS = {
  AI_LEVEL_RND: 'ai_level_rnd', // Always random
  AI_LEVEL_KI: 'ai_level_ki', // Killer instinct, random until a winning move is found
  AI_LEVEL_HAH: 'ai_level_hah', // Half and half, sometimes SMRT and sometimes KI
  AI_LEVEL_SMRT: 'ai_level_smrt', // Use past games to learn how to win or draw
};

const symbols = [X, O];
const sides = [PLAYER, CPU];
const coinSides = ['heads', 'tails'];
const initOpenCells = [1, 2, 3, 4, 5, 6, 7, 8, 9];
const winningPatterns = [
  [1, 2, 3], // TOP ROW
  [4, 5, 6], // MIDDLE ROW
  [7, 8, 9], // BOTTOM ROW
  [1, 4, 7], // LEFT COL
  [2, 5, 8], // MIDDLE COL
  [3, 6, 9], // RIGHT COL
  [1, 5, 9], // TL-BR DIAG
  [3, 5, 7], // TR-BL DIAG
];

const LOCAL_STORAGE_KEYS = {
  HISTORY: 'tic-tac-know-history',
  WINS_LOSSES: 'tic-tac-know-wins-losses',
};

const whoseTurnMessages = {
  player: 'Your turn',
  cpu: 'CPU Turn',
};
const gameOverMessages = {
  player: 'YOU WIN!',
  cpu: 'CPU WINS',
  draw: 'The Game is a draw.',
};

const APP_STATES = {
  GAME: 'game',
  MENU: 'menu',
};

const MENU_STATUSES = {
  PICK_SIDE: 'pick-side',
  PICK_AI: 'pick-ai',
  START: 'start',
  COMPLETE: 'complete',
};

const sideMessages = {
  x: 'You are playing as X. The CPU is playing as O.',
  o: 'You are playing as O. The CPU is playing as X.',
};
const aiLevelLabels = {
  ai_level_rnd: 'Herp Derp',
  ai_level_ki: 'Killer Instinct',
  ai_level_hah: 'Half & Half',
  ai_level_smrt: 'Mastermind',
};
const aiLevelMessages = {
  ai_level_rnd:
    'Your oppnent is, to put it bluntly, not terribly bright, making moves (and inapropriate noises) at random.',
  ai_level_ki:
    'Your opponent knows enough to be annoyingly good, but just like all your friends from college, it cannot plan ahead.',
  ai_level_hah:
    'Your opponent is playing with one chip tied behind its back, using a mix of strategies, sometimes smart and sometimes not.',
  ai_level_smrt:
    "Your opponent is smarter than you are, more disciplined and, let's face it, better looking. You don't stand a chance.",
};

//
export {
  symbols,
  sides,
  coinSides,
  PLAYER,
  CPU,
  X,
  O,
  whoseTurnMessages,
  gameOverMessages,
  sideMessages,
  aiLevelMessages,
  aiLevelLabels,
  initOpenCells,
  winningPatterns,
  AI_LEVELS,
  LOCAL_STORAGE_KEYS,
  APP_STATES,
  MENU_STATUSES,
};
