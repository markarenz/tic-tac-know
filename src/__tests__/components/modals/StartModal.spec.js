import { render, fireEvent, screen, waitFor } from '@testing-library/react';
import { act } from 'react-dom/test-utils';
import StartModal from '../../../components/modals/StartModal';
import { AI_LEVELS, X, O } from '../../../constants';

const mocks = {
  show: true,
  handleStartClick: jest.fn(),
  handleResetClick: jest.fn(),
};
describe('StartModal - X', () => {
  it('renders component - PLAY Click', async () => {
    const container = render(
      <StartModal
        show={mocks.show}
        handleStartClick={mocks.handleStartClick}
        handleResetClick={mocks.handleResetClick}
      />
    );
    await act(async () => {
      fireEvent(
        screen.getByTestId('startmenu-play'),
        new MouseEvent('click', {
          bubbles: true,
          cancelable: true,
        })
      );
    });
    await new Promise((resolve) => setTimeout(resolve, 0));
    await waitFor(() => {
      expect(container).toMatchSnapshot();
    });
  });
});
