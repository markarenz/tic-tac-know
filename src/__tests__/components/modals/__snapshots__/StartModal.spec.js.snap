// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`StartModal - X renders component - PLAY Click 1`] = `
Object {
  "asFragment": [Function],
  "baseElement": <body>
    <div>
      <div
        class="text-center"
        data-testid="transition-wrap"
      >
        <div
          class="text-center mt-16 inline-block"
        >
          <div
            class="relative"
          >
            <div
              class="animate-pingSlowSm absolute top-0 left-0 inline-flex h-full w-full rounded-lg bg-purple-800 opacity-100"
            />
            <div
              class="mb-8"
            >
              <button
                class="relative px-16 py-4 text-4xl border-2 border-purple-300 text-white bg-gradient-to-t from-black to-purple-700 rounded-lg scale-100 hover:scale-[1.2] transition-scale font-bold duration-300"
                data-testid="startmenu-play"
              >
                PLAY
              </button>
            </div>
          </div>
          <div
            class="mt-4"
          >
            <button
              class="relative px-4 py-2 text-lg border-2 border-purple-300 text-white bg-gradient-to-t from-black to-gray-500 rounded-lg scale-100 hover:scale-[1.2] transition-scale duration-300"
              data-testid="startmenu-reset"
            >
              RESET
            </button>
          </div>
        </div>
      </div>
    </div>
  </body>,
  "container": <div>
    <div
      class="text-center"
      data-testid="transition-wrap"
    >
      <div
        class="text-center mt-16 inline-block"
      >
        <div
          class="relative"
        >
          <div
            class="animate-pingSlowSm absolute top-0 left-0 inline-flex h-full w-full rounded-lg bg-purple-800 opacity-100"
          />
          <div
            class="mb-8"
          >
            <button
              class="relative px-16 py-4 text-4xl border-2 border-purple-300 text-white bg-gradient-to-t from-black to-purple-700 rounded-lg scale-100 hover:scale-[1.2] transition-scale font-bold duration-300"
              data-testid="startmenu-play"
            >
              PLAY
            </button>
          </div>
        </div>
        <div
          class="mt-4"
        >
          <button
            class="relative px-4 py-2 text-lg border-2 border-purple-300 text-white bg-gradient-to-t from-black to-gray-500 rounded-lg scale-100 hover:scale-[1.2] transition-scale duration-300"
            data-testid="startmenu-reset"
          >
            RESET
          </button>
        </div>
      </div>
    </div>
  </div>,
  "debug": [Function],
  "findAllByAltText": [Function],
  "findAllByDisplayValue": [Function],
  "findAllByLabelText": [Function],
  "findAllByPlaceholderText": [Function],
  "findAllByRole": [Function],
  "findAllByTestId": [Function],
  "findAllByText": [Function],
  "findAllByTitle": [Function],
  "findByAltText": [Function],
  "findByDisplayValue": [Function],
  "findByLabelText": [Function],
  "findByPlaceholderText": [Function],
  "findByRole": [Function],
  "findByTestId": [Function],
  "findByText": [Function],
  "findByTitle": [Function],
  "getAllByAltText": [Function],
  "getAllByDisplayValue": [Function],
  "getAllByLabelText": [Function],
  "getAllByPlaceholderText": [Function],
  "getAllByRole": [Function],
  "getAllByTestId": [Function],
  "getAllByText": [Function],
  "getAllByTitle": [Function],
  "getByAltText": [Function],
  "getByDisplayValue": [Function],
  "getByLabelText": [Function],
  "getByPlaceholderText": [Function],
  "getByRole": [Function],
  "getByTestId": [Function],
  "getByText": [Function],
  "getByTitle": [Function],
  "queryAllByAltText": [Function],
  "queryAllByDisplayValue": [Function],
  "queryAllByLabelText": [Function],
  "queryAllByPlaceholderText": [Function],
  "queryAllByRole": [Function],
  "queryAllByTestId": [Function],
  "queryAllByText": [Function],
  "queryAllByTitle": [Function],
  "queryByAltText": [Function],
  "queryByDisplayValue": [Function],
  "queryByLabelText": [Function],
  "queryByPlaceholderText": [Function],
  "queryByRole": [Function],
  "queryByTestId": [Function],
  "queryByText": [Function],
  "queryByTitle": [Function],
  "rerender": [Function],
  "unmount": [Function],
}
`;
